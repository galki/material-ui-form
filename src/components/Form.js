// @flow

import React from 'react'
import _ from 'lodash'

import FormControl from '@material-ui/core/FormControl'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import FormHelperText from '@material-ui/core/FormHelperText'
import FormLabel from '@material-ui/core/FormLabel'
import InputLabel from '@material-ui/core/InputLabel'
import Checkbox from '@material-ui/core/Checkbox'
import Switch from '@material-ui/core/Switch'

import FormControlClone from './FormControlClone'
import FormControlLabelClone from './FormControlLabelClone'
import FieldClone from './FieldClone'
import CheckableFieldClone from './CheckableFieldClone'
import DeleteFieldRowButton from './DeleteFieldRowButton'
import propNames from '../propNames'
import {
  messageMap,
  validate,
  validators as defaultValidators,
  constants as validationConstants,
} from '../validation'


function verifyFieldElement(component: any): boolean {
  const whitelist = [
    FormControlLabel,
  ]

  return whitelist.includes(component.type)
   || (_.has(component, 'props.name') && _.has(component, 'props.value'))
}

function extractFieldValidators(fieldProps: Object): Array<mixed> {
  let validators = _.get(fieldProps, propNames.FIELD_VALIDATORS)
  if (validators !== undefined) {
    if (_.isString(validators)) {
      validators = validators.replace(/\s/g, '').split(',')
    } else if (!_.isArray(validators)) {
      validators = [validators]
    }
    return validators
  }
  return []
}

function getFieldValues(fields: Object): Object {
  const values = {}
  _.each(fields, (field, name) => {
    if (_.get(field, 'checked') !== false) {
      values[name] = field.value
    }
  })
  return values
}

function getPristineFieldValues(fields: Object): Object {
  const values = {}
  _.each(fields, (field, name) => {
    if (!field.isPristine && _.get(field, 'checked') !== false) {
      values[name] = field.pristineValue
    }
  })
  return values
}

function getFieldTemplate() {
  return {
    isDirty: false,
    isPristine: true,
    isRequired: null,
    pristineValue: null,
    step: undefined,
    validations: [],
    validators: [],
    value: undefined,
  }
}

function deriveErrorSteps(fields: Object): Array<number> {
  const errorSteps = []
  _.each(fields, (field) => {
    if (field.validations.length > 0 && !errorSteps.includes(field.step)) {
      errorSteps.push(field.step)
    }
  })
  return errorSteps
}

function isValidForm(fields: Object): boolean {
  return _.size(_.filter(fields, field => field.validations.length > 0)) === 0
}

type Props = {
  activeStep?: number,
  autoComplete?: string,
  children: Array<mixed>,
  className?: Object,
  disableSubmitButtonOnError?: boolean,
  onFieldValidation?: Function,
  onSubmit: Function,
  onValuesChange?: void | Function,
  style?: Object,
  validation?: {
    messageMap?: Object,
    messageMapKeyPrefix?: string,
    requiredValidatorName?: string | boolean,
    validators?: Object,
    validate?: Function,
    validateInputOnBlur?: boolean,
  },
  validations: Object,
  id?: Object,
  method?: Object,
  action?: Object,
  name?: Object,
};

type State = {
  disableSubmitButton: boolean,
  fields: Object,
};

export default class Form extends React.Component<Props, State> {
  static defaultProps = {
    activeStep: 0,
    autoComplete: 'off',
    className: undefined,
    disableSubmitButtonOnError: true,
    onFieldValidation: undefined,
    onValuesChange: undefined,
    style: {},
    validation: {},
    validations: {},
    id: undefined,
    method: undefined,
    action: undefined,
    name: undefined,
  }

  // eslint-disable-next-line react/sort-comp
  onValuesChange: void

  constructor(props: Object) {
    super(props)

    this.onValuesChange = props.onValuesChange
    this.validation = Object.assign(this.validation, props.validation)
    this.state = {
      disableSubmitButton: false,
      fields: {},
    }
  }

  static getDerivedStateFromProps(nextProps: Object, prevState: Object) {
    const { fields } = prevState

    if (!_.isEmpty(fields)) {
      // add validations to fields
      _.each(nextProps.validations, (validations, name) => {
        if (_.has(fields, name)) {
          fields[name].validations = validations
        } else {
          // eslint-disable-next-line no-console
          console.warn(`validations field "${name}" does not exist`)
        }
      })
      return { fields }
    }
    return null
  }

  validation = {
    messageMap,
    messageMapKeyPrefix: '',
    requiredValidatorName: validationConstants.REQUIRED_VALIDATOR_NAME,
    validators: defaultValidators,
    validate,
    validateInputOnBlur: false,
  }

  onFieldConstruct = (fieldProps: Object) => {
    const {
      checked,
      name,
      required,
      value,
    } = fieldProps

    // checkable input
    if (checked === true) {
      _.defer(() => {
        this.setState({
          fields: {
            ...this.state.fields,
            [name]: {
              ...getFieldTemplate(),
              checked: checked || false,
              step: this.props.activeStep,
              value,
            },
          },
        })
      })
    // other inputs
    } else if (!_.isBoolean(checked)) {
      const { requiredValidatorName } = this.validation
      if (!_.has(this.state.fields, name)) {
        const validators = extractFieldValidators(fieldProps)

        if (required && !_.isEmpty(requiredValidatorName)) {
          validators.unshift(requiredValidatorName)
        }
        const isRequired = required || validators.includes(requiredValidatorName)
        // set any validations on first construct
        let validations = []
        if (!_.has(this.state.fields, name)
          && _.has(this.props.validations, name)
        ) {
          validations = this.props.validations[name]
        }

        _.defer(() => {
          this.setState({
            fields: {
              ...this.state.fields,
              [name]: {
                ...getFieldTemplate(),
                isRequired,
                pristineValue: value,
                step: this.props.activeStep,
                validators,
                validations,
                value,
              },
            },
          })

          if (!_.isEmpty(value)) {
            this.validateField(name, value)
          }
        })
      }
    }
  }

  onFieldValueChange = (name: string, value: any, isDirty: boolean = false) => {
    _.defer(() => {
      this.setState({
        fields: {
          ...this.state.fields,
          [name]: {
            ...this.state.fields[name],
            isDirty: isDirty || this.state.fields[name].isDirty,
            isPristine: false,
            validations: [],
            value,
          },
        },
      })

      if (isValidForm(this.state.fields)) {
        this.enableSubmitButton()
      }
      if (this.onValuesChange !== undefined) {
        this.onValuesChange(
          getFieldValues(this.state.fields),
          getPristineFieldValues(this.state.fields)
        )
      }

      if (this.state.fields[name].isDirty) {
        this.validateField(name, value)
      }
    })
  }

  onFieldToggle = (name: string, value: any, checked: boolean) => {
    this.setState({
      fields: {
        ...this.state.fields,
        [name]: {
          ...this.state.fields[name],
          checked,
          isPristine: false,
          validations: [],
          value,
        },
      },
    })
  }

  validateField = (name: string, value: any) => {
    const field = this.state.fields[name]

    if (!(field.value === '' && !field.isRequired) && !_.isEmpty(field.validators)) {
      const { validation } = this
      const validations = validation.validate(value, field.validators, validation)

      // update state
      field.validations = validations
      this.setState({
        fields: {
          ...this.state.fields,
          [name]: field,
        },
      })
      // disable submit button
      if (!_.isEmpty(validations)) {
        this.disableSubmitButton()
      }
      // propogate validation
      if (this.props.onFieldValidation !== undefined) {
        let errorSteps
        if (field.step !== undefined) {
          errorSteps = deriveErrorSteps(this.state.fields)
        }
        this.props.onFieldValidation(field, errorSteps)
      }
    }
  }

  reset = () => {
    const { fields } = this.state
    _.defer(() => {
      _.each(fields, (field, name) => {
        this.setState({
          fields: {
            ...this.state.fields,
            [name]: {
              ...this.state.fields[name],
              isDirty: false,
              isPristine: true,
              value: '',
            },
          },
        })
      })
    })
  }

  submit = (event: Event) => {
    event.preventDefault()
    let isValid = true
    const { fields } = this.state
    _.each(fields, (field, name) => {
      if (field.isRequired && field.value === '') {
        this.validateField(name, '')
        isValid = false
      }
    })
    if (isValid) {
      this.props.onSubmit(
        getFieldValues(fields),
        getPristineFieldValues(fields)
      )
    }
  }

  enableSubmitButton() {
    if (this.state.disableSubmitButton) {
      this.setState({ disableSubmitButton: false })
    }
  }

  disableSubmitButton() {
    if (this.props.disableSubmitButtonOnError) {
      this.setState({ disableSubmitButton: true })
    }
  }

  deleteRow = (row: string) => {
    const pos: number = row.indexOf('[')
    const rowName: string = row.substr(0, pos)
    const rowIndex: number = parseInt(row.substr(pos + 1), 10)

    const { fields } = this.state
    _.each(fields, (field, fieldName) => {
      if (fieldName.startsWith(row)) {
        delete fields[fieldName]
      } else if (fieldName.startsWith(rowName)) {
        const index = parseInt(fieldName.substr(pos + 1), 10)
        if (index > rowIndex) {
          const newRow = fieldName.replace(/\[\d+\]/, `[${(index - 1)}]`)
          delete fields[fieldName]
          fields[newRow] = field
        }
      }
    })

    this.setState({ fields })
  }

  cloneChildrenRecursively(children: any): any {
    return React.Children.map(children, (child) => {
      if (_.isEmpty(child)) {
        return null
      }
      if (_.isString(child)) {
        return child
      }

      const isFieldElement = verifyFieldElement(child)
      const nestedChildren = _.isArray(child.props.children) && !isFieldElement
        ? _.filter(child.props.children, v => (_.isObject(v) || _.isString(v)))
        : false

      // nested elements
      if (nestedChildren !== false) {
        // FormControl element with field/group name-value props
        if (child.type === FormControl) {
          const fieldElement = nestedChildren.find(el =>
            ![FormLabel, InputLabel, FormHelperText].includes(el.type)
            && el.props.name !== undefined
            && el.props.value !== undefined)
          if (fieldElement !== undefined) {
            const { name } = fieldElement.props
            return (
              <FormControlClone
                key={name}
                field={this.state.fields[name]}
                formControlComp={child}
                onConstruct={this.onFieldConstruct}
                onValueChange={this.onFieldValueChange}
              />
            )
          }
        }
        // non-FormControl element
        return (
          React.cloneElement(child, {
            children: this.cloneChildrenRecursively(nestedChildren),
          })
        )
      }
      // add disable functionality to submit button
      if (child.props.type === 'submit') {
        return React.cloneElement(child, {
          disabled: this.state.disableSubmitButton,
        })
      // non-interactive elements should be rendered as is
      } else if (!isFieldElement) {
        // delete row button
        if (child.props[propNames.DELETE_FIELD_ROW] !== undefined) {
          return (<DeleteFieldRowButton
            buttonComp={child}
            onRequestRowDelete={this.deleteRow}
          />)
        }
        // any other element
        return child
      }
      // clone control label
      if (child.type === FormControlLabel) {
        const { name } = child.props.control.props
        return (
          <FormControlLabelClone
            key={name}
            field={this.state.fields[name]}
            control={child.props.control}
            label={child.props.label}
            onConstruct={this.onFieldConstruct}
            onToggle={this.onFieldToggle}
          />
        )
      }
      // clone input element
      const { name } = child.props

      // checkable
      if (child.type === Checkbox || child.type === Switch) {
        return (
          <CheckableFieldClone
            key={name}
            field={this.state.fields[name]}
            fieldComp={child}
            onConstruct={this.onFieldConstruct}
            onToggle={this.onFieldToggle}
          />
        )
      }

      return (
        <FieldClone
          key={name}
          field={this.state.fields[name]}
          fieldComp={child}
          onConstruct={this.onFieldConstruct}
          onValueChange={this.onFieldValueChange}
          useNativeRequiredValidator={!this.validation.requiredValidatorName}
          validateInputOnBlur={this.validation.validateInputOnBlur}
        />
      )
    })
  }

  render() {
    return (
      <form
        autoComplete={this.props.autoComplete}
        className={this.props.className}
        onReset={this.reset}
        onSubmit={this.submit}
        style={this.props.style}
        id={this.props.id}
        method={this.props.method}
        action={this.props.action}
        name={this.props.name}
      >
        { this.cloneChildrenRecursively(this.props.children) }
      </form>
    )
  }
}
