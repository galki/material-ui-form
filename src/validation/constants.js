export default {
  REQUIRED_VALIDATOR_NAME: 'isRequired',
}
