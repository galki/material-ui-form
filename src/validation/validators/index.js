import validators from 'validator' // eslint-disable-line import/no-extraneous-dependencies

import isAlias from './isAlias'
import isDate from './isDate'
import isNumber from './isNumber'
import isSerial from './isSerial'
import isTime from './isTime'
import isSize from './isSize'
import isRequired from './isRequired'


validators.isAlias = isAlias
validators.isDate = isDate
validators.isNumber = isNumber
validators.isSerial = isSerial
validators.isTime = isTime
validators.isSize = isSize
validators.isRequired = isRequired

export default validators
