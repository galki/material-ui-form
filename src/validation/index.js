import _ from 'lodash' // eslint-disable-line import/no-extraneous-dependencies

import messageMap from './messageMap'
import validators from './validators'

export constants from './constants'


function sprintf(str, args) {
  let predicate
  if (_.isString(args)) {
    predicate = args
  } else if (_.isObject(args) && !_.isArray(args)) {
    args = Object.values(args)
    predicate = (match, number) => (
      args[number] !== undefined ? args[number] : match
    )
  } else {
    predicate = (match, number) => (
      // eslint-disable-next-line no-nested-ternary
      args[number] !== undefined
        ? (_.isArray(args) ? args.join(', ') : args[number])
        : match
    )
  }
  return str.replace(/{(\d+)}/g, predicate)
}

let validationMessageMap = _.clone(messageMap)

export const createValidation = (validatorName, args, config) => {
  if (!_.isEmpty(config.messageMap)) {
    validationMessageMap = config.messageMap
  }

  let code = validatorName
  // first check if prefix code exists
  if (!validatorName.startsWith(config.messageMapKeyPrefix)) {
    const prefixedCode = `${config.messageMapKeyPrefix}${validatorName}`
    if (_.has(validationMessageMap, prefixedCode)) {
      code = prefixedCode
    }
  }

  let message = validationMessageMap[code]
  if (message !== undefined && (_.isNumber(args) || !_.isEmpty(args))) {
    message = sprintf(message, args)
  }
  return { code, message }
}

export const validate = (value, fieldValidators, config) => {
  const validations = []
  if (_.isEmpty(fieldValidators)) {
    return []
  } else if (!_.isArray(fieldValidators)) {
    // eslint-disable-next-line no-console
    console.error('invalid validators format:', fieldValidators)
    return false
  }

  fieldValidators.forEach((validator) => {
    let args
    let validatorName = validator
    if (_.isObject(validator) && _.size(validator) === 1) {
      args = Object.values(validator)[0] // eslint-disable-line prefer-destructuring
      validatorName = Object.keys(validator)[0] // eslint-disable-line prefer-destructuring
    } else if (!_.isString(validator)) {
      console.error('invalid validator:', validator) // eslint-disable-line
    }

    if (config.validators[validatorName] === undefined) {
      console.error('undefined validator:', validatorName) // eslint-disable-line
    } else {
      value = String(value)
      const validation = config.validators[validatorName](value, args)
      if (!validation) {
        validations.push(createValidation(validatorName, args, config))
      }
    }
  })

  return validations
}

export {
  messageMap,
  validators,
}
