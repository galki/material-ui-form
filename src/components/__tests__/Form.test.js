import React from 'react'
import { shallow } from 'enzyme' // eslint-disable-line import/no-extraneous-dependencies
import toJson from 'enzyme-to-json' // eslint-disable-line import/no-extraneous-dependencies

import Button from 'material-ui/Button'
import TextField from 'material-ui/TextField'
import Checkbox from 'material-ui/Checkbox'
import { FormControl, FormControlLabel } from 'material-ui/Form'
import { InputLabel } from 'material-ui/Input'
import Select from 'material-ui/Select'
import { MenuItem } from 'material-ui/Menu'

import Form from '../Form'
import { messageMap, validators } from '../../validation'


/* eslint-disable react/self-closing-comp */

describe('<Form>', () => {
  validators.isBorat = value => value === 'borat'
  const customMessageMap = Object.assign(messageMap, {
    isBorat: 'NAAAAAT! You can only write "borat" lol',
  })
  const validation = {
    messageMap: customMessageMap,
    requiredValidatorName: false,
    validators,
  }
  const validations = { name: [{ code: 'isAlpha', message: 'invalid' }] }

  const wrapper = shallow(
    <Form
      autoComplete="on"
      disableSubmitButtonOnError={false}
      onValuesChange={jest.fn()}
      onSubmit={jest.fn()}
      validation={validation}
      validations={validations}
    >
      {'fill out the form!'}
      <TextField
        label="Name"
        type="text"
        name="name"
        value="123"
        data-validators="isRequired,isBorat"
      />
      <div>
        <Checkbox checked={false} name="foo" value="bar" />
        <span>FooBar</span>
        <FormControlLabel
          control={<Checkbox checked={false} name="love" value="yes" />}
          label="I love love"
        />
        <React.Fragment></React.Fragment>
      </div>
      <FormControl required>
        <InputLabel>Age</InputLabel>
        <Select value="" name="age">
          <MenuItem value=""><em>Please select your age ...</em></MenuItem>
          <MenuItem value={10}>Teens</MenuItem>
          <MenuItem value={20}>Twenties</MenuItem>
          <MenuItem value={30}>Thirties</MenuItem>
          <MenuItem value="40+">Fourties +</MenuItem>
        </Select>
      </FormControl>
      <Button variant="raised" type="submit">Submit</Button>
    </Form>
  )

  it('should render', () => {
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should have set validation', () => {
    expect(wrapper.instance().validation).toHaveProperty('messageMap')
  })

  it('should set state', () => {
    expect(wrapper.state()).toMatchObject({
      disableSubmitButton: false,
      fields: {},
    })
  })

  it('should have 2 children', () => {
    expect(wrapper.children()).toHaveLength(5)
  })

  // it('should have a cloned TextField', () => {
  //   expect(wrapper.find(TextField)).toBeInstanceOf(FieldClone)
  // })

  // it('should have a non-cloned Button', () => {
  //   expect(wrapper.find(Button)).not.toBeInstanceOf(FieldClone)
  // })

  it('should handle submit event', () => {
    const { onSubmit } = wrapper.instance().props
    const event = new Event('submit')
    wrapper.simulate('submit', event)
    expect(onSubmit).toHaveBeenCalled()
  })

  it('should have submit button disabled', () => {
    expect(wrapper.instance().props.disableSubmitButtonOnError).toEqual(false)
    expect(wrapper.state('disableSubmitButton')).toEqual(false)
  })
})
