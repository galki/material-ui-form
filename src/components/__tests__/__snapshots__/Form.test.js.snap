// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form> (all props) should render 1`] = `
<form
  autoComplete="on"
  onReset={[Function]}
  onSubmit={[Function]}
>
  fill out the form!
  <FieldClone
    field={Object {}}
    fieldComp={
      <TextField
        data-validators="isRequired,isBorat"
        label="Name"
        name="name"
        required={false}
        select={false}
        type="text"
        value="123"
      />
    }
    key="name/.1"
    onConstruct={[Function]}
    onValueChange={[Function]}
    useNativeRequiredValidator={true}
    validateInputOnBlur={false}
  />
  <div
    key=".2"
  >
    <CheckableFieldClone
      field={Object {}}
      fieldComp={
        <WithStyles(Checkbox)
          checked={false}
          name="foo"
          value="bar"
        />
      }
      key="foo/.0"
      onConstruct={[Function]}
      onToggle={[Function]}
    />
    <span
      key=".1"
    >
      FooBar
    </span>
    <FormControlLabelClone
      control={
        <WithStyles(Checkbox)
          checked={false}
          name="love"
          value="yes"
        />
      }
      field={Object {}}
      key="love/.2"
      label="I love love"
      onConstruct={[Function]}
      onToggle={[Function]}
    />
    <React.Fragment
      key=".3"
    />
  </div>
  <FormControlClone
    field={Object {}}
    formControlComp={
      <WithStyles(FormControl)
        required={true}
      >
        <WithStyles(InputLabel)>
          Age
        </WithStyles(InputLabel)>
        <WithStyles(Select)
          name="age"
          value=""
        >
          <WithStyles(MenuItem)
            value=""
          >
            <em>
              Please select your age ...
            </em>
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={10}
          >
            Teens
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={20}
          >
            Twenties
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={30}
          >
            Thirties
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value="40+"
          >
            Fourties +
          </WithStyles(MenuItem)>
        </WithStyles(Select)>
      </WithStyles(FormControl)>
    }
    key="age/.3"
    onConstruct={[Function]}
    onValueChange={[Function]}
  />
  <WithStyles(Button)
    disabled={false}
    key=".4"
    type="submit"
    variant="raised"
  >
    Submit
  </WithStyles(Button)>
</form>
`;

exports[`<Form> (no props) should render 1`] = `
<form
  autoComplete="off"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <input
    key=".0"
    type="text"
  />
</form>
`;
